### PR TITLE
Enable any protocol support for redirects

### LINKED ISSUE #1245: Redirects: Support URLs with other protocols such as `mailto:` or `ipfs://`
### Description

Currently, redirects only support URLs of type `http`, `https` and local routes.

I propose adding support to those of type `mailto` as well.

Discussion about the topic started on https://github.com/vikejs/vike/discussions/1244.

### Motivation
- `mailto` URLs are the next more common type of emails. Some people might want to setup a certain route to a `mailto`.
- For example, in order to be easily contactable, I might setup a route in my personal website called `/mail-me` and having it redirect to my current email, as opposed to giving my email to a person (which might be a complex email prone to error when typing) I can simply give an easy to memorize link
- On top of that, `mailto` urls also support a bunch of parameters such as setting the default body and subject — which might be useful for certain use-cases, such as job applications, etc.

### PR DESCRIPTION
**Pull request by usk94**
resolves #1245

This PR introduces the capability to utilize the mailto protocol within the redirect configuration. 
This allows for configured path to be redirected to email addresses, triggering the opening of the user's default email client. 

config example
```javascript
// vite.config.js

import vike from 'vike/plugin'

export default {
  plugins: [
    vike({
      redirects: {
        '/email': 'mailto:foo@bar.test'
      }
    })
  ]
}
```

### CODE REVIEW

File: vike/shared/route/resolveRedirects.ts [hunk 1]
```diff
@@ -41,9 +42,11 @@ function resolveRouteStringRedirect(urlSource: string, urlTarget: string, urlPat
     }
     urlResolved = urlResolved.replaceAll(key, val)
   })
-  assert(!urlResolved.includes('@'))
+  if (!urlResolved.startsWith('mailto:')) {
+    assert(!urlResolved.includes('@'), 'URL should not contain "@" unless it is a mailto link.')
+  }
   if (urlResolved === urlPathname) return null
-  assert(urlTarget.startsWith('/') || urlTarget.startsWith('http'))
+  assert(urlTarget.startsWith('/') || urlTarget.startsWith('http') || urlResolved.startsWith('mailto:'))
```
Comment (brillout) [inline vike/shared/route/resolveRedirects.ts:49]:
```suggestion
  assertUsage(urlTarget.startsWith('/') || urlTarget.startsWith('http') || urlResolved.startsWith('mailto:'))
```

File: vike/shared/route/resolveRedirects.ts [hunk 2]
```diff
@@ -41,9 +42,11 @@ function resolveRouteStringRedirect(urlSource: string, urlTarget: string, urlPat
     }
     urlResolved = urlResolved.replaceAll(key, val)
   })
-  assert(!urlResolved.includes('@'))
+  if (!urlResolved.startsWith('mailto:')) {
+    assert(!urlResolved.includes('@'), 'URL should not contain "@" unless it is a mailto link.')
```
Comment (brillout) [inline vike/shared/route/resolveRedirects.ts:46]:
```suggestion
    assertUsage(!urlResolved.includes('@'), 'URL should not contain "@" unless it is a mailto link.')
```

File: vike/shared/route/resolveRedirects.ts [hunk 3]
```diff
@@ -41,9 +42,11 @@ function resolveRouteStringRedirect(urlSource: string, urlTarget: string, urlPat
...
+  }
   if (urlResolved === urlPathname) return null
-  assert(urlTarget.startsWith('/') || urlTarget.startsWith('http'))
+  assertUsage(urlTarget.startsWith('/') || urlTarget.startsWith('http') || urlResolved.startsWith('mailto:'))
```
Comment (brillout) [inline vike/shared/route/resolveRedirects.ts:49]:
```suggestion
  assert(urlTarget.startsWith('/') || urlTarget.startsWith('http') || urlResolved.startsWith('mailto:'))
```
I suggested the wrong assertion.

Comment (usk94) [reply vike/shared/route/resolveRedirects.ts:49]: 🙆‍♂️

**Comment by usk94** [comment]
@brillout 
> I wonder if this could be extended to support any other protocol such as ipfs:, ipns:, magnet:, etc. If all protocols match /^[a-zA-Z]:/, then I guess we could simply replace the http:// checks with that regex.

I wonder `javascript:` and `data:` are at risk of xss vulnerability, so how about validating them?https://cheatsheetseries.owasp.org/cheatsheets/XSS_Filter_Evasion_Cheat_Sheet.html

**Comment by brillout** [comment]
> I wonder `javascript:` and `data:` are at risk of xss vulnerability, so how about validating them?https://cheatsheetseries.owasp.org/cheatsheets/XSS_Filter_Evasion_Cheat_Sheet.html

What kind of vulnerability do you have in mind? Do you see a problem with setting the `Location` header to any arbitrary string? I could be wrong, but I'd assume browsers to already validate and not do anything funky upon arbitrary `Location` values.

As for https://cheatsheetseries.owasp.org/cheatsheets/Unvalidated_Redirects_and_Forwards_Cheat_Sheet.html#dangerous-url-redirect-example-1 it don't think it applies to us: it's something the Vike user needs to careful about and I don't see how Vike could implement additional protections.

**Comment by usk94** [comment]
@brillout 
Understood. 
I was concerned more about the possibility of malicious developers exploiting end users rather than xss vulnerabilities. 
But you're right, that's something developers need to be cautious about, and there's nothing we can do about it on our end👍
Thank you for pointing that out!

**Comment by brillout** [comment]
👍 That said, I could very well miss out a problematic situation, but so far I can't think of one. I guess it's ok if we risk it.

**Comment by usk94** [comment]
I have switched to using regex🙆‍♂️

**Comment by brillout** [comment]
I did some polishing, everything's perfect IMO :) Do we agree and shall we merge?

**Comment by usk94** [comment]
It's fine. Thank you for your reviewing and refactoring!

File: vike/utils/parseUrl-extras.ts [hunk 4]
```diff
@@ -109,3 +110,8 @@ function addUrlOrigin(url: string, origin: string | null): string {
   const urlModified = createUrlFromComponents(origin, pathnameOriginal, searchOriginal, hashOriginal)
   return urlModified
 }
+
+function isUriWithProtocol(uri: string): boolean {
+  // Is it correct to assume that all protocols (http:, https:, ipfs:, magnet:, ...) match this RegExp?
```
Comment (ndavd) [inline vike/utils/parseUrl-extras.ts:115]: Albeit not common, URL schemes may contain digits, `+`, `.` and `-`.

**Comment by ndavd** [comment]
Added a small comment @brillout

**Comment by ndavd** [comment]
> Do you have an example?

https://en.wikipedia.org/wiki/List_of_URI_schemes

**Comment by brillout** [comment]
Neat. Fixed: 72ede8a12e5482405590656f79a2219374d0213c.

**Comment by ndavd** [comment]
Looks great 🔥

**Pull request merged**